work/worker.go

```diff
@@ -0,0 +1,151 @@
+package work
+
+import (
+	"context"
+	"sync"
+
+	"github.com/filecoin-project/lotus/node/modules/helpers"
+	logging "github.com/ipfs/go-log/v2"
+	"go.uber.org/fx"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/sentinel-visor/schedule"
+)
+
+var log = logging.Logger("worker-manager")
+
+func NewWorkerManager(mctx helpers.MetricsCtx, lc fx.Lifecycle) *WorkerManager {
+	return &WorkerManager{
+		WorkerID: 1,
+		Workers:  make(map[int]Worker),
+	}
+}
+
+// WorkerManager controls Workers allowing them to be started and stopped independently of each other.
+type WorkerManager struct {
+	WorkerID  int
+	Workers   map[int]Worker
+	WorkersMu sync.Mutex
+}
+
+type Worker interface {
+	Start(ctx context.Context)
+	Stop(ctx context.Context)
+	Err() error
+	Running() bool
+}
+
+// AddTaskAsWorker wraps schedule.Task `t` as a TaskWorker, adds it to WorkerManager's Worker list,
+// and returns its ID. Workers can be started and stopped using their ID's.
+func (wm *WorkerManager) AddTaskAsWorker(t schedule.Task) int {
+	tw := &TaskWorker{
+		task:    t,
+		running: false,
+	}
+	return wm.AddWorker(tw)
+}
+
+// AddWorker adds Worker `w` to WorkerManager's Worker list and returns its ID. Workers can be
+// started and stopped using their ID's.
+func (wm *WorkerManager) AddWorker(w Worker) int {
+	wm.WorkersMu.Lock()
+	defer func() {
+		wm.Workers[wm.WorkerID] = w
+		wm.WorkerID++
+		wm.WorkersMu.Unlock()
+	}()
+	log.Infow("added worker", "id", wm.WorkerID)
+	return wm.WorkerID
+}
+
+// Start starts the worker with ID `id`, errors if the ID is not found, or id the worker us running.
+func (wm *WorkerManager) Start(ctx context.Context, id int) error {
+	wm.WorkersMu.Lock()
+	defer wm.WorkersMu.Unlock()
+
+	w, ok := wm.Workers[id]
+	if !ok {
+		return xerrors.Errorf("starting worker ID: %d not found", id)
+	}
+	if w.Running() {
+		return xerrors.Errorf("starting worker ID: %d already running", id)
+	}
+	log.Infow("starting worker", "id", id)
+	w.Start(ctx)
+	return nil
+}
+
+// Stop stops the worker with ID `id`, errors if the ID is not found or if the worker is not running.
+func (wm *WorkerManager) Stop(ctx context.Context, id int) error {
+	wm.WorkersMu.Lock()
+	defer wm.WorkersMu.Unlock()
+
+	w, ok := wm.Workers[id]
+	if !ok {
+		return xerrors.Errorf("stopping worker ID: %d not found", id)
+	}
+	if !w.Running() {
+		return xerrors.Errorf("stopping worker ID: %d is not running", id)
+	}
+	log.Infow("stopping worker", "id", id)
+	w.Stop(ctx)
```

Comment:
Should there be a timeout here? 

Should we allow `w.Stop` to return an error?